### DISCUSSION TITLE
Clarification on required types/adding multiple requirements

### DISCUSSION THREAD

**Discussion by JHunnicutt** [joe-bell/cva#157] [Q&A]
I'm still lost on how to properly use required variants even after consulting the [docs](https://cva.style/docs/getting-started#required-variants). Below I have a component that I would like to make the intent and size required. I have optional empty, because I have no values that I want to be optional at this time.

Currently, I'm able to include the two values listed one level deep after `required`, namely `intent`, and `size`. I can't however find how to structure the code to list the values for `intent` and `size`, which would be `primary` and the two size values. I'm also lost how to include multiple required items in `cvaButton()`, as anything more than what I currently have gives me type errors.

```
import { AriaButtonProps, useButton } from 'react-aria'
import { useRef } from 'react'
import { cva, type VariantProps } from 'class-variance-authority'

export type ButtonVariantProps = VariantProps<typeof cvaButton>

const cvaButton = cva('button rounded-full uppercase', {
	variants: {
		required: {
			intent: {
				primary: [
					'px-8',
					'py-4',
					'text-base',
					'font-semibold',
					'tracking-widest',
					'text-white',
				],
			},
			size: {
				large: 'w-56',
				fullWidth: 'w-full',
			},
		},
		optional: {},
	},
})

type requiredButtonProps = Required<Pick<ButtonVariantProps, 'required'>>
type optionalButtonProps = Pick<ButtonVariantProps, 'optional'>

interface ButtonProps extends AriaButtonProps, requiredButtonProps, optionalButtonProps {}

export const Button: React.FC<ButtonProps> = ({ required, ...props }: ButtonProps) => {
	let ref = useRef(null)
	let { buttonProps } = useButton(props, ref)
	let { children } = props

	return (
		<button
			{...buttonProps}
			ref={ref}
			className={cvaButton({ required: 'intent' })}
		>
			{children}
		</button>
	)
}
```

Here is the component being used. Again, I'm confused because TypeScript is not providing me to access `primary` or `size`.
```
<Button required={'intent'}>Test button</Button>
```

Thanks!

**Comment by joe-bell** [comment]
Hey! I think maybe there's some confusion here – perhaps something to improve in the docs.

`cva` doesn't provide a required/optional API for variants. The example you've linked to is purely illustrative of how a variant (named `required`) could have required types and vice versa

So in your case you'll want to remove the variants nested within the `optional`/`required` keys (this isn't supported) and pick out the variants you want to be required via the `Required<>` helper

**Comment by JHunnicutt** [comment]
Thanks for the reply!

Your comment was enough to jog me out of incorrect thinking. Instead of following along exactly by structuring `required` and `optional`, I just used the TypeScript pattern you explained to make my `intent` and `size` be required. I can't believe I went down the road as long as I did before making that realization. Such is the way of development!
